src/java/main/org/apache/zookeeper/ClientCnxn.java

```diff
@@ -1080,6 +1080,8 @@ private void startConnect() throws IOException {
                         zooKeeperSaslClient.shutdown();
                     }
                     zooKeeperSaslClient = new ZooKeeperSaslClient(getServerPrincipal(addr), clientConfig);
+                    // SASL login succeeded
+                    saslLoginFailed = false;
```

Comment:
this change has impact on tunnelAuthInProgress.  But yes, we should init the variable on new connection start as this is tunnelAuthInProgress logic expects.